hypothesis_bio/strategies.py

```diff
@@ -149,3 +150,196 @@ def sequence(draw, *, alphabet: Alphabet = Alphabet.DNA, max_size: int = 1000):
+    AGA="R",
+    AGC="S",
+    AGG="R",
+    AGT="S",
+    ATA="I",
+    ATC="I",
+    ATG="M",
+    ATT="I",
+    CAA="Q",
+    CAC="H",
+    CAG="Q",
+    CAT="H",
+    CCA="P",
+    CCC="P",
+    CCG="P",
+    CCT="P",
+    CGA="R",
+    CGC="R",
+    CGG="R",
+    CGT="R",
+    CTA="L",
+    CTC="L",
+    CTG="L",
+    CTT="L",
+    GAA="E",
+    GAC="D",
+    GAG="E",
+    GAT="D",
+    GCA="A",
+    GCC="A",
+    GCG="A",
+    GCT="A",
+    GGA="G",
+    GGC="G",
+    GGG="G",
+    GGT="G",
+    GTA="V",
+    GTC="V",
+    GTG="V",
+    GTT="V",
+    TAA="*",
+    TAC="Y",
+    TAG="*",
+    TAT="Y",
+    TCA="S",
+    TCC="S",
+    TCG="S",
+    TCT="S",
+    TGA="*",
+    TGC="C",
+    TGG="W",
+    TGT="C",
+    TTA="L",
+    TTC="F",
+    TTG="L",
+    TTT="F",
+)
+
+ALL_CODONS = tuple(CANONICAL_GENE_CODE.keys())
+CANONICAL_STOP_CODONS = ("TAA", "TGA", "TAG")
+
+
+@st.composite
+def stop_codon(draw, stop_codons=CANONICAL_STOP_CODONS):
+    """
+    A :mod:`hypothesis` strategy for building stop codons
+
+    Parameters
+    ----------
+    draw
+        For internal hypothesis use.
+
+    stop_codons: list[string]
+            a list of stop codons (defaults to the canonical TAA, TGA, and TAG)
+
+    Returns
+    -------
+    string
+        a single stop codon
+    ----------
+    """
+    return draw(st.sampled_from(stop_codons))
+
+
+@st.composite
+def codon(draw, *, stop_codons=CANONICAL_STOP_CODONS):
+    """
+    A :mod:`hypothesis` strategy for getting one codon
+
+    Parameters
+    ----------
+    draw
+        For internal hypothesis use.
+
+    stop_codons: list[string]
+        The list of accepted stop codons
+
+    Returns
+    -------
+    string
+        a singel codon
+    ----------
+    """
+    return draw(st.sampled_from(ALL_CODONS))
+
+
+@st.composite
+def non_stop_codon(draw, *, stop_codons=CANONICAL_STOP_CODONS):
+    """
+    A :mod:`hypothesis` strategy for building non-stop codons
+
+    Parameters
+    ----------
+    draw
+        For internal hypothesis use.
+
+    stop_codons: list[string]
+        The list of accepted stop codons
+
+    **kwargs
+        Additional arguments sent to :mod:`lists`
+
+    Returns
+    -------
+    string
+        a single non-stop codon
+    ----------
+    """
+    cdn = draw(codon())
+    assume(cdn not in stop_codons)
+    return cdn
+
+
+@st.composite
+def coding_sequence(
+    draw,
+    *,
+    include_stop_codon=True,
+    include_start_codon=True,
+    allow_internal_stops=False,
+    **kwargs
+):
+    """
+    A :mod:`hypothesis` strategy for building coding sequences
+
+    Parameters
+    ----------
+    draw
+        For internal hypothesis use.
+
+    include_stop_codon: bool
+        If True, append a stop codon
+
+    include_start_codon: bool
+        If True, append a start codon (ATG)
+
+    allow_internal_stops: bool
+        If True, allow stop codons inside the codings sequence
+
+    **kwargs
+        Additional arguments passed to :mod:`codons`
```

Comment:
They're actually passed to `st.lists()`...

And it's possible to define this one entirely without `@st.composite` too, if you think about it carefully:

```python
return st.builds(
    "{}{}{}".format,
    st.just("ATG" if include_start_codon else ""),
    st.lists(codon() if include_stop_codon else non_stop_codon(), **kwargs).map("".join),
    stop_codon() if include_stop_codon else st.just(""),
)
```

Plus some carefully juggling of `**kwargs`` for size if you want it to be correct with respect to the possible start and end codons.

Reply:
Thanks for the review @Zac-HD ! I'm pushing a couple more commits that I believe address everything you mention.